code/modules/chemistry/tools/robospray.dm

```diff
@@ -18,29 +17,48 @@
 	)
 	var/currentreagent = "epinephrine"
 	var/propername = "Epinephrine"
+	var/image/fluid_image
+	var/last_refill = -1
+	var/extra_refill = 0
 
 	hide_attack = 2
 	inventory_counter_enabled = 1
 
 	New()
 		..()
 		processing_items.Add(src)
+		src.update_icon()
 
 	disposing()
 		..()
 		processing_items.Remove(src)
 
+	proc/update_icon()
+		if (botreagents[currentreagent] >= 1)
+			if (!src.fluid_image)
+				src.fluid_image = image(src.icon, "hypoover", -1)
+			var/datum/reagent/R = reagents_cache[src.currentreagent]
+			src.fluid_image.color = rgb(R.fluid_r, R.fluid_g, R.fluid_b, 255)
+			src.UpdateOverlays(src.fluid_image, "fluid")
+		else
+			src.UpdateOverlays(null, "fluid")
+		src.inventory_counter.update_number(botreagents[currentreagent])
+		signal_event("icon_updated")
+
 	attack_self(mob/user as mob)
-		picker = picker % length(botreagents) + 1
-		currentreagent = botreagents[picker]
-		var/datum/reagent/temp_reagent = reagents_cache[currentreagent]
-		propername = temp_reagent.name
+		var/available_chems = list()
+		for (var/reagent in botreagents)
+			available_chems += reagents_cache[reagent]
+		var/datum/reagent/pick = input(usr, "Inject which chemical?", "Cybernetic Hypospray", null) in available_chems
+		currentreagent = pick.id
+		propername = pick.name
 		user.show_text("[src] is now injecting [propername], [botreagents[currentreagent]] units left.", "blue")
+		update_icon()
 		tooltip_rebuild = 1
 		return
```

Comment:
> I believe `input` will prevent execution of the later bits here until a selection has been made, which might mess with things if your module is removed in the meantime or you drop the tool etc.
> Likely want to check for `src` existence after `input`, and arguably want to make sure it's still within `user`'s inventory.

both true

Reply:
I looked at the portable seed fabricator and it doesn't have any special logic for if the item gets removed while the input box is open. Since this doesn't have any effect outside of the item itself, the worst that can happen if the module gets removed is that a cyborg can minorly inconvenience another cyborg one time after going through a lot of convoluted steps.

Reply:
Still a bug.

Reply:
Do you have an example of an item that does the check properly?

Reply:
Not off the top of my head, but consider something like:
```suggestion
		var/curr_loc = src.loc
		var/datum/reagent/pick = input(usr, "Inject which chemical?", "Cybernetic Hypospray", null) in available_chems
		if (src && user && src.loc == curr_loc)
			currentreagent = pick.id
			propername = pick.name
			user.show_text("[src] is now injecting [propername], [botreagents[currentreagent]] units left.", "blue")
			update_icon()
			tooltip_rebuild = 1
		return
```
Note that I have not tested this.

Reply:
I can't find any cyborg items that check for the item being removed after calling `input`.

Reply:
Alright, turns out none of them had the sanity check. Added it to all of them (unless I missed something).